File: julia/Euler/src/bin/p0081.jl

# project euler: problem 81

module Prob0081

function solve_0081(fname::String = "0081_matrix.txt", fn::Function = min)
    data = vcat(([parse(Int, x) for x in split(l, ",")]' for l in readlines(joinpath((@__DIR__), "../../assets", fname)))...)

    data[1, :] = accumulate(+, data[1, :])
    data[:, 1] = accumulate(+, data[:, 1])
    for i = axes(data, 1)[begin + 1:end]
        for j = axes(data, 2)[begin + 1:end]
            data[i, j] += fn(data[i - 1, j], data[i, j - 1])
        end
    end
    data[end, end]
end

end #module

using .Prob0081: solve_0081
export solve_0081

File: rust/src/bin/p0081.rs
// Project Euler: Problem 81

euler::run_solver!(81);

static FILE_DATA: &str = include_str!("../../assets/0081_matrix.txt");

fn solve() -> String {
    compute(FILE_DATA).to_string()
}

fn compute(data: &str) -> u64 {
    use std::cmp;

    let mut matrix = parse_data(data);
    let mut prev: Vec<_> = matrix[0]
        .iter()
        .scan(0, |acc, x| {
            *acc += *x;
            Some(*acc)
        })
        .collect();
    prev.insert(0, u64::MAX);

    for work in &mut matrix[1..] {
        (*work).insert(0, u64::MAX);
        for i in 1..((*work).len()) {
            work[i] += cmp::min(work[i - 1], prev[i]);
        }
        prev.clone_from(work);
    }

    *prev.last().unwrap()
}

fn parse_data(data: &str) -> Vec<Vec<u64>> {
    let mut ret: Vec<Vec<u64>> = Vec::new();

    for line in data.lines() {
        ret.push(line.split(',').map(|s| s.parse::<u64>().unwrap()).collect());
    }

    ret
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0081() {
        assert_eq!(compute(super::FILE_DATA), 427337);
    }
}
